### PR TITLE
Add Simple AES Class 

### PR DESCRIPTION
**Pull request by AlirezaP**
## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by Ellerbach** [comment]
Can you please add a .gitignore in the root of the project? So it will be easier to review and won't bring the binaries in the commits?

File: System.Security.Cryptography/System.Security.Cryptography/AES.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
+﻿using System;
+
+namespace System.Security.Cryptography
+{
+    public class AES
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/AES.cs:5]: you are missing the intellisense comments everywhere

File: System.Security.Cryptography/System.Security.Cryptography/AES.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
+﻿using System;
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/AES.cs:1]: please add the header

File: System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("CSharp.TestApplication")]
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs:8]: please adjust this file appropriately and don't forget about the header

File: System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//
+//      Major Version
+//      Minor Version 
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs:32]: that's not needed

File: System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+[assembly: AssemblyFileVersion("1.0.0.0")]
+
+/////////////////////////////////////////////////////////////////
+// This attribute is mandatory when building Interop libraries //
+// update this whenever the native assembly signature changes  //
+[assembly: AssemblyNativeVersion("1.0.0.0")]
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/Properties/AssemblyInfo.cs:38]: that's not needed neither except if you have native elements which doesn't seems to be the case

File: System.Security.Cryptography/System.Security.Cryptography/packages.config
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/packages.config:1]: you need a package lock as well

File: System.Security.Cryptography/System.Security.Cryptography/System.Security.Cryptography.nfproj [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectTypeGuids>{11A8DD76-328B-46DF-9F39-F559912D0360};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
+    <ProjectGuid>cf405d7f-957a-40d9-becf-62849e6b8a9e</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <FileAlignment>512</FileAlignment>
+    <RootNamespace>System.Security.Cryptography</RootNamespace>
+    <AssemblyName>System.Security.Cryptography</AssemblyName>
+    <TargetFrameworkVersion>v1.0</TargetFrameworkVersion>
+  </PropertyGroup>
```
Comment (Ellerbach) [inline System.Security.Cryptography/System.Security.Cryptography/System.Security.Cryptography.nfproj:18]: please add all what's needed to sign the project

**Review by Ellerbach — COMMENTED** [review]
Thanks, couple of remarks to start, once with a gitignore, it's going to be more clear :-)
Also, you're missing all the nuget and related elements including the pipeline. Checkout the docs: https://docs.nanoframework.net/content/maintainers/creating-a-new-repo.html#prepare-the-initial-commit

**Comment by josesimoes** [comment]
This is a bit early in the process. The "base" project is still pending review and merge. Only after that it makes sense to start adding new features. 😉

**Comment by AlirezaP** [comment]
so am I supposed to continue or close the Pull request ?

**Comment by josesimoes** [comment]
> so am I supposed to continue or close the Pull request ?

Please hold until the inital version is there. Then you can merge on your working branch and take it from there.
BTW, we're following the .NET API on these. Please make sure to follow AES API from [here](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography?view=net-7.0&WT.mc_id=DT-MVP-5004179).

**Comment by AlirezaP** [comment]
sure. I would be happy if I could participate in this work

**Comment by AlirezaP** [comment]
> Can you please add a .gitignore in the root of the project? So it will be easier to review and won't bring the binaries in the commits?

sure. i will do that

File: System.Security.Cryptography/System.Security.Cryptography/AES.cs (hunk 2 above)
Comment (AlirezaP) [reply System.Security.Cryptography/System.Security.Cryptography/AES.cs]: If I understand correctly ("add the header"). i done

**Comment by Ellerbach** [comment]
@AlirezaP can you please make sure that you don't commit the .vs folder and other non needed folders? Once done, it will be better and easier to review the code. Thanks

**Comment by josesimoes** [comment]
@AlirezaP just took a quick look at this. Can you please point me to the .NET system.security.cryptography API that has this AES class? The policy is to follow the API from the full framework whenever possible.

**Comment by AlirezaP** [comment]
> @AlirezaP can you please make sure that you don't commit the .vs folder and other non needed folders? Once done, it will be better and easier to review the code. Thanks

@Ellerbach I obeyed the mentioned note. As a test, I made a change in the .vs folder (I added a text file); it doesn't included in the git change.

**Comment by Ellerbach** [comment]
> > @AlirezaP can you please make sure that you don't commit the .vs folder and other non needed folders? Once done, it will be better and easier to review the code. Thanks
> 
> @Ellerbach Laurent Ellerbach FTE I obeyed the mentioned note. As a test, I made a change in the .vs folder (I added a text file); it doesn't included in the git change.

You have to get rid of all the files I4ve highlighted because they are part of your PR:
image

There are 2 options:
* make sure you remove them from your next commit
* Copy the AES.cs and nfproj somewhere else on your hard drive, close the PR, start clean from the main banch, clean the folders, create a brnahc bring back the 2 files and raise a new PR

**Comment by josesimoes** [comment]
The main goal here is to rely on the native provider for all these. For AES the full .NET version is going back and fort simply because it seems that AES has shown to have some security issues... or at least some versions(implementations) of it...

I understand (and agree) that going through the MemoryStream and CryptoStream is a bit too much for embedded systems scenarios. This seems to be pointing to a need to have a lightweight approach to this API as we've done a few times.
Which will happen (most likely) with other algorithms. 

This may sound a blocker, still: we should come up with a list of algorithms that we want to have implemented and find out a common denominator for the methods and API.

**Comment by Ellerbach** [comment]
> This may sound a blocker, still: we should come up with a list of algorithms that we want to have implemented and find out a common denominator for the methods and API.

I agree and on the simplification as well. We've been added some of the crypto algo first purely on a managed way because we mainly we needed them to connect to cloud providers like Azure and AWS.

We've always been also very pragmatic with all this. And we never were afraid to refactor some elements when we were able to cluster those. So, we should not neither be too afraid of bringing couple of those at first and refactor a bit later.

**Comment by AlirezaP** [comment]
> The main goal here is to rely on the native provider for all these. For AES the full .NET version is going back and fort simply because it seems that AES has shown to have some security issues... or at least some versions(implementations) of it...
> 
> I understand (and agree) that going through the MemoryStream and CryptoStream is a bit too much for embedded systems scenarios. This seems to be pointing to a need to have a lightweight approach to this API as we've done a few times. Which will happen (most likely) with other algorithms.
> 
> This may sound a blocker, still: we should come up with a list of algorithms that we want to have implemented and find out a common denominator for the methods and API.

you are right. however, it is worth noting that the existence of a basic implementation of this algorithm in the Nanoframeworks security library can be useful; it can be customed and optimized proportional to the embedded devices in the future.

**Comment by AlirezaP** [comment]
> > > @AlirezaP can you please make sure that you don't commit the .vs folder and other non needed folders? Once done, it will be better and easier to review the code. Thanks
> > 
> > 
> > @Ellerbach Laurent Ellerbach FTE I obeyed the mentioned note. As a test, I made a change in the .vs folder (I added a text file); it doesn't included in the git change.
> 
> You have to get rid of all the files I4ve highlighted because they are part of your PR: image
> 
> There are 2 options:
> 
> * make sure you remove them from your next commit
> * Copy the AES.cs and nfproj somewhere else on your hard drive, close the PR, start clean from the main banch, clean the folders, create a brnahc bring back the 2 files and raise a new PR

Ok, I will remove them from the next commit.

**Comment by AlirezaP** [comment]
> > > > @AlirezaP can you please make sure that you don't commit the .vs folder and other non needed folders? Once done, it will be better and easier to review the code. Thanks
> > > 
> > > 
> > > @Ellerbach Laurent Ellerbach FTE I obeyed the mentioned note. As a test, I made a change in the .vs folder (I added a text file); it doesn't included in the git change.
> > 
> > 
> > You have to get rid of all the files I4ve highlighted because they are part of your PR: image
> > There are 2 options:
> > 
> > * make sure you remove them from your next commit
> > * Copy the AES.cs and nfproj somewhere else on your hard drive, close the PR, start clean from the main banch, clean the folders, create a brnahc bring back the 2 files and raise a new PR
> 
> Ok, I will remove them from the next commit.

Actually, I clone the branch in a new project, and everything seems to be ok.

**Comment by Ellerbach** [comment]
> Actually, I clone the branch in a new project, and everything seems to be ok.

It is not. Can you please either remove all the files, either start clean?
I'm not going to merge a PR containing non necessary files. You have the list of files in the PR, github shows it.
Thank you.

**Comment by AlirezaP** [comment]
@Ellerbach excuse me I didn't understand before. :)
it's done

File: System.Security.Cryptography/AES.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+using System;
+
+namespace System.Security.Cryptography
+{
+    /// <summary>
+    /// Represents the class for aes encryption/decryption
+    /// </summary>
+    public class AES
+    {
+        public enum EncryptionModes { ECB }
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:15]: Is this aligned with the Full .NET ?
And is this enum also part of the class or outside of the class?
In all cases, you have to add intellisesne comments on it

File: System.Security.Cryptography/AES.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        /// </summary>
+        /// <param name="key">The secret key for XOR opration with block </param>
+        /// <param name="block">The Block of data for XOR opration with secret key</param>
+        /// <exception cref="ArgumentException"></exception>
+        private void DecryptBlock(byte[] key, byte[] block)
+        {
+            // Ensure that the key and block have the same length
+            if (key.Length != block.Length)
+            {
+                throw new ArgumentException("Key and block must have the same length.");
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:153]: please move the text of the exception in the intellisense comments, so we save space in the device ;-)

File: System.Security.Cryptography/AES.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            for (int i = 0; i < block.Length; i++)
+            {
+                block[i] = (byte)(block[i] ^ key[i]);
+            }
+        }
+
+        /// <summary>
+        /// Decrypt the array of bytes In ECB Mode
+        /// </summary>
+        /// <param name="key"></param>
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:111]: you are missing the intellisense comments. If it's a private function, you don't have to add intelisense. But if you add it, please make sure it's complete.

File: System.Security.Cryptography/AES.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        /// </summary>
+        /// <param name="key">The secret key for XOR opration with block </param>
+        /// <param name="block">The Block of data for XOR opration with secret key</param>
+        /// <exception cref="ArgumentException"></exception>
+        private void EncryptBlock(byte[] key, byte[] block)
+        {
+            // Ensure that the key and block have the same length
+            if (key.Length != block.Length)
+            {
+                throw new ArgumentException("Key and block must have the same length.");
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:99]: move also the text into the intellisense comment, so we gain space in the device.

File: System.Security.Cryptography/AES.cs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+            return buf;
+        }
+
+        /// <summary>
+        /// Encrypt the array of bytes In ECB Mode
+        /// </summary>
+        /// <param name="key">The secret key to use for the symmetric algorithm.</param>
+        /// <param name="data">Array of bytes for encryption</param>
+        /// <returns></returns>
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:59]: same remark here, even if it's a private function and even if intellisense comments are not mandatory, if you add them, please add them properly.

File: System.Security.Cryptography/AES.cs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+            return buf;
+        }
+
+        /// <summary>
+        /// Decrypt the array of bytes.
+        /// </summary>
+        /// <param name="key">The secret key to use for the symmetric algorithm.</param>
+        /// <param name="data">encrypted array of bytes for decryption</param>
+        /// <returns></returns>
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:41]: You are missing the comments for the return type.

File: System.Security.Cryptography/AES.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        public enum EncryptionModes { ECB }
+        public EncryptionModes Mode { get; set; } = EncryptionModes.ECB;
+
+        /// <summary>
+        /// Encrypt the array of bytes.
+        /// </summary>
+        /// <param name="key">The secret key to use for the symmetric algorithm.</param>
+        /// <param name="data">array of bytes for encryption</param>
+        /// <returns></returns>
```
Comment (Ellerbach) [inline System.Security.Cryptography/AES.cs:23]: you are missing the comments for the return type

File: System.Security.Cryptography/AES.cs (hunk 7 above)
Comment (AlirezaP) [reply System.Security.Cryptography/AES.cs]: I made some changes here.
the Full .Net uses this "Represents the abstract base class from which all implementations of the Advanced Encryption Standard (AES) must inherit." but this implementation doesn't use abstract class.
should we set excactly Full .Net comments here?

Comment (Ellerbach) [reply System.Security.Cryptography/AES.cs]: >should we set excactly Full .Net comments here?

No, it's ok :-) Important is mainly the definitions of the functions and the intellisense comments associated to them.

What about the enum?

**Review by Ellerbach — COMMENTED** [review]
looks good all up. Can you add an example of usage in the main README?

**Comment by AlirezaP** [comment]
> looks good all up. Can you add an example of usage in the main README?

sure. Done

File: README.md [hunk 14]
```diff
@@ -33,7 +33,36 @@ byte[] hash = hmacsha256.ComputeHash(Encoding.UTF8.GetBytes(encodedUri + "\n" +
...
+### AES
 
+Advanced Encryption Standard (AES)
+
+This version supports only the ECB mode
+
+The following example demonstrates how to encrypt and decrypt sample data by using the Aes class.
+
+```csharp
+            //Sample Usage
```
Comment (Ellerbach) [inline README.md:45]: can you please have all this left shifted?

**Review by Ellerbach — COMMENTED** [review]
Now we are good with the code, would be perfect to add few tests. You can get inspiration from the existing ones. This will allow to ensure it's still working fine. Few cases should be enough.

File: Tests/System.Security.CryptographyTests/AESTests.cs [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace System.Security.CryptographyTests
+{
+    [TestClass]
+    public class AESTests
+    {
+        static byte[] cipherDataArray = new byte[] { 112, 15, 93, 166, 173, 66, 95, 251, 63, 172, 69, 69, 182, 109, 13, 93 };
+        static byte[] clearDataArray = new byte[] { 78, 97, 110, 111, 102, 114, 97, 109, 101, 119, 111, 114, 107, 0, 0, 0 };
+        static byte[] key = new byte[16] { 62, 110, 51, 201, 203, 48, 62, 150, 90, 219, 42, 55, 221, 109, 13, 93 };
+
```
Comment (Ellerbach) [inline Tests/System.Security.CryptographyTests/AESTests.cs:18]:
```suggestion
```

File: Tests/System.Security.CryptographyTests/AESTests.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+            // Encrypt the bytes
+            var enData = aes.Encrypt(key, clearTextByteArrayWithPadding);
+            CollectionAssert.AreEqual(cipherDataArray,enData);
+
+
+            // Decrypt the bytes
+            var decryptedByteArray = aes.Decrypt(enData, key);
+            CollectionAssert.AreEqual(clearDataArray, decryptedByteArray);
+
```
Comment (Ellerbach) [inline Tests/System.Security.CryptographyTests/AESTests.cs:38]:
```suggestion
```

**Review by Ellerbach — COMMENTED** [review]
2 little extra lines. Then good for me. @josesimoes does this PR change the native signature?

**Pull request merged**